LANGUAGE_SDK_SPECIFICATION.md

```diff
@@ -0,0 +1,1325 @@
+- **Level 2 (Standard)**: All required operations, serialization, logging
+- **Level 3 (Complete)**: All operations, concurrency, advanced patterns
+
+## 17. Versioning and Compatibility
+
+### 17.1 Specification Versioning
+
+This specification uses semantic versioning:
+
+- Major: Breaking changes to required behavior
+- Minor: New optional features
+- Patch: Clarifications and corrections
+
+### 17.2 Backward Compatibility
+
+SDKs SHOULD:
+
+1. Maintain backward compatibility within major versions
+2. Provide migration guides for breaking changes
+3. Support gradual adoption of new features
+
+## 18. References
+
+### 18.1 AWS Lambda Documentation
+
+- Lambda Durable Functions: https://docs.aws.amazon.com/lambda/latest/dg/durable-functions.html
+- Lambda API Reference: https://docs.aws.amazon.com/lambda/latest/api/
+- Lambda Runtime API: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html
+
+### 18.2 Related Standards
+
+- RFC 2119: Key words for use in RFCs to Indicate Requirement Levels
+- JSON: ECMA-404, RFC 8259
+- ISO 8601: Date and time format
+
+## Appendix A: Example State Transitions
+
+### A.1 Simple Step Execution
+
+```
+Invocation 1:
+  - Load state: []
+  - Start STEP(id="step1")
+  - Checkpoint: START step1
+  - Execute user function
+  - Checkpoint: SUCCEED step1, payload="result"
+  - Return: SUCCEEDED, Result="result"
+
+Execution completes successfully.
+```
+
+### A.2 Step with Retry
+
+```
+Invocation 1:
+  - Load state: []
+  - Start STEP(id="step1")
+  - Checkpoint: START step1
+  - Execute user function (fails)
+  - Checkpoint: RETRY step1, NextAttemptDelaySeconds=5
+  - Return: PENDING
+
+Invocation 2 (after 5 seconds):
+  - Load state: [step1: PENDING, Attempt=1]
+  - Resume STEP(id="step1")
+  - Checkpoint: START step1 (attempt 2)
+  - Execute user function (succeeds)
+  - Checkpoint: SUCCEED step1, payload="result"
+  - Return: SUCCEEDED, Result="result"
+
+Execution completes successfully.
+```
+
+### A.3 Wait Operation
+
+```
+Invocation 1:
+  - Load state: []
+  - Start WAIT(id="wait1", seconds=60)
+  - Checkpoint: START wait1, WaitSeconds=60
+  - Return: PENDING
+
+Invocation 2 (after 60 seconds):
+  - Load state: [wait1: SUCCEEDED]
+  - Skip WAIT(id="wait1") - already completed
+  - Continue execution
+  - Return: SUCCEEDED, Result="done"
+
+Execution completes successfully.
+```
+
+### A.4 Replay After Lambda Timeout
+
+```
+Invocation 1:
+  - Load state: []
+  - Start STEP(id="step1")
+  - Checkpoint: START step1
+  - Execute user function (succeeds)
+  - Checkpoint: SUCCEED step1, payload="result1"
+  - Start STEP(id="step2")
+  - Checkpoint: START step2
+  - Execute user function (long-running)
+  - [Lambda timeout occurs]
+
+Invocation 2 (Lambda retry):
+  - Load state: [step1: SUCCEEDED, step2: STARTED]
+  - Replay STEP(id="step1") - return cached "result1"
+  - Resume STEP(id="step2")
+  - Checkpoint: START step2 (same ID, continues)
```

Comment:
You wouldn't checkpoint START again - it's already started. Depends on semantics but can either run it again then checkpoint success/failure/retry or decide to immediately checkpoint failure, or retry, etc.